hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -815,17 +805,33 @@ def cached_test_function(self, buffer):
         Otherwise we call through to ``test_function``, and return a
         fresh result.
         """
-        result = self.tree.lookup(buffer)
-        if result is not None:
+        buffer = hbytes(buffer)
+        try:
+            return self.__data_cache[buffer]
+        except KeyError:
+            pass
+
+        rewritten, status = self.tree.rewrite(buffer)
+
+        try:
+            result = self.__data_cache[rewritten]
+        except KeyError:
+            pass
+        else:
+            self.__data_cache[buffer] = result
             return result
 
         # We didn't find a match in the tree, so we need to run the test
         # function normally. Note that test_function will automatically
         # add this to the tree so we don't need to update the cache.
-        result = ConjectureData.for_buffer(buffer)
-        self.test_function(result)
-        if result.status == Status.OVERRUN:
-            return Overrun
+        if status == Status.OVERRUN:
+            result = Overrun
+        else:
+            result = ConjectureData.for_buffer(buffer)
+            self.test_function(result)
+            if result.status == Status.OVERRUN:
+                result = Overrun
```

Comment:
My instinct here is to combine the two `Overrun` replacements into one:

```python
if status != Status.OVERRUN:
    result = ConjectureData.for_buffer(buffer)
    self.test_function(result)
    # TODO Maybe emit a diagnostic if the predicted status doesn't match
    status = result.status

if status == Status.OVERRUN:
    # Either the tree predicted an overrun, or the test function produced an overrun
    result = Overrun
```

But I don't know if that's actually an improvement, and it does give worse branch coverage, so 🤷🏻‍♂️.